lib/util/getopt/getopt.c

```diff
@@ -39,7 +39,93 @@ LOG_MODULE_REGISTER(getopt);
 #define	BADARG	((int)':')
 #define	EMSG	""
 
-void getopt_init(struct getopt_state *state)
+
+#if CONFIG_GETOPT_SINGLE
+int opterr = 1;	/* if error message should be printed */
+int optind = 1;	/* index into parent argv vector */
+int optopt;	/* character checked for validity */
+int optreset;	/* reset getopt */
+char *optarg;	/* argument associated with option */
+
+static char *m_place = EMSG;
+#if CONFIG_GETOPT_LONG_SINGLE
+static int m_nonopt_start;
+static int m_nonopt_end;
+#endif /* CONFIG_GETOPT_LONG_SINGLE */
+
+static void state_store(struct z_getopt_state *state)
+{
+	state->opterr = opterr;
+	state->optind = optind;
+	state->optopt = optopt;
+	state->optreset = optreset;
+	state->optarg = optarg;
+	state->place = m_place;
+
+#if CONFIG_GETOPT_LONG
+	state->nonopt_start = m_nonopt_start;
+	state->nonopt_end = m_nonopt_end;
+#endif
+}
+
+static void state_restore(struct z_getopt_state *state)
+{
+	opterr = state->opterr;
+	optind = state->optind;
+	optopt = state->optopt;
+	optreset = state->optreset;
+	optarg = state->optarg;
+	m_place = state->place;
+
+#if CONFIG_GETOPT_LONG
+	m_nonopt_start = state->nonopt_start;
+	m_nonopt_end = state->nonopt_end;
+#endif
+}
+
+int getopt(int nargc, char *const nargv[], const char *ostr)
+{
+	struct z_getopt_state state;
+	int ret;
+
+	state_store(&state);
+	ret = z_getopt(&state, nargc, nargv, ostr);
```

Comment:
why not using global `struct z_getopt_state` all the time without need for store and restore?

Reply:
Unfortunately we can't do it this way because the normal use of the getopt function requires global variables.